README
GStreamer test program for testing elements.

Will take source video from the videotestsrc or from a video file, run it through elements of you choice, including overlays etc. and then output video to the screen in a standalone window via autovideosink.

Build:
	$ ./autogen
	$ make

Run:
	$ cd src
	$ ./gst-app



Legacy readme:
Based on the gst-app template for writing your own GStreamer-based app.

The code is deliberately kept simple so that you quickly understand the basics
of how to set up autotools and your source tree.

This template demonstrates :
- what to do in autogen.sh
- how to setup configure.ac (your package name and version, GStreamer flags)
- how to setup your source dir 
  - main () and main header
  - supporting source code
  - Makefile.am

The template performs one simple function : loading a pipeline from xml and
start to iterate it.  Try saving a pipeline from gst-editor, it will play
with this sample program.

More features might get added to this template later on.


ChangeLog
See revision history.


src/main.c
/* Copyright (C) 2006 Tim-Philipp Müller <tim centricular net>
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 * Alternatively, the contents of this file may be used under the
 * GNU Lesser General Public License Version 2.1 (the "LGPL"), in
 * which case the following provisions apply instead of the ones
 * mentioned above:
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "gst-app.h"

static void
handle_file_or_directory (const gchar * filename)
{
  GError *err = NULL;
  GDir *dir;
  gchar *uri;

  if ((dir = g_dir_open (filename, 0, NULL))) {
    const gchar *entry;

    while ((entry = g_dir_read_name (dir))) {
      gchar *path;

      path = g_strconcat (filename, G_DIR_SEPARATOR_S, entry, NULL);
      handle_file_or_directory (path);
      g_free (path);
    }

    g_dir_close (dir);
    return;
  }

  if (g_path_is_absolute (filename)) {
    uri = g_filename_to_uri (filename, NULL, &err);
  } else {
    gchar *curdir, *absolute_path;

    curdir = g_get_current_dir ();
    absolute_path = g_strconcat ( curdir, G_DIR_SEPARATOR_S, filename, NULL);
    uri = g_filename_to_uri (absolute_path, NULL, &err);
    g_free (absolute_path);
    g_free (curdir);
  }

  if (uri) {
    /* great, we have a proper file:// URI, let's play it! */
    play_uri (uri);
  } else {
    g_warning ("Failed to convert filename '%s' to URI: %s", filename,
        err->message);
    g_error_free (err);
  }

  g_free (uri);
}

int
main (int argc, char *argv[])
{

	  // Before calling any GLib or GStreamer function, we must initialise
	  // the GLib threading system
	  if (!g_thread_supported())
	    g_thread_init (NULL);

	  gst_init(&argc, &argv);

	  // For this test, just go straight to playing some pipeline
	  play_camera();


/*
  gchar **filenames = NULL;
  const GOptionEntry entries[] = {
    // you can add your own command line options here
    { G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_FILENAME_ARRAY, &filenames,
      "Special option that collects any remaining arguments for us" },
    { NULL, }
  };
  GOptionContext *ctx;
  GError *err = NULL;
  gint i, num;

  // Before calling any GLib or GStreamer function, we must initialise
  // the GLib threading system
  if (!g_thread_supported())
    g_thread_init (NULL);

  ctx = g_option_context_new ("[FILE1] [FILE2] ...");
  g_option_context_add_group (ctx, gst_init_get_option_group ());
  g_option_context_add_main_entries (ctx, entries, NULL);

  if (!g_option_context_parse (ctx, &argc, &argv, &err)) {
    g_print ("Error initializing: %s\n", GST_STR_NULL (err->message));
    return -1;
  }
  g_option_context_free (ctx);

  if (filenames == NULL || *filenames == NULL) {
    g_print ("Please specify a file to play\n\n");
    return -1;
  }



  num = g_strv_length (filenames);

  for (i = 0; i < num; ++i) {
    handle_file_or_directory (filenames[i]);
  }

  g_strfreev (filenames);
*/
  return 0;
}


src/play.h
/* Copyright (C) 2006 Tim-Philipp Müller <tim centricular net>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 * Alternatively, the contents of this file may be used under the
 * GNU Lesser General Public License Version 2.1 (the "LGPL"), in
 * which case the following provisions apply instead of the ones
 * mentioned above:
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef _MY_APP_PLAY_H_INCLUDED_
#define _MY_APP_PLAY_H_INCLUDED_

#include <gst/gst.h>

void    play_uri (const gchar * uri);
void    play_camera(void);

#endif /* _MY_APP_PLAY_H_INCLUDED_ */



src/play.c
/* Copyright (C) 2016 Paul Barber
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 * Alternatively, the contents of this file may be used under the
 * GNU Lesser General Public License Version 2.1 (the "LGPL"), in
 * which case the following provisions apply instead of the ones
 * mentioned above:
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include "play.h"
//#include <unistd.h> // for usleep

static GMainLoop *loop;

static gboolean
my_bus_callback (GstBus *bus, GstMessage *message, gpointer data)
{
	g_print("Got %s message from %s\n", GST_MESSAGE_TYPE_NAME(message), GST_MESSAGE_SRC_NAME(message));

	switch (GST_MESSAGE_TYPE(message)) {
	case GST_MESSAGE_ERROR: {
		GError *err;
		gchar *debug;

		gst_message_parse_error (message, &err, &debug);
		g_print ("Error: %s (%s)\n", err->message, debug);
		g_error_free(err);
		g_free(debug);

		g_main_loop_quit(loop);
		break;
	}
	case GST_MESSAGE_EOS:
		g_main_loop_quit(loop);
		break;
	default:
		break;
	}

	return TRUE;  // continue to watch for messages
}


void
play_camera(void)
{
	GstElement *pipeline;
	GstElement *source;
	GstElement *sink;
	GstBus *bus;
	guint bus_watch_id;

	pipeline = gst_pipeline_new("VideoPipeline");

	g_print ("Trying to create source ... ");
	source = gst_element_factory_make("videotestsrc", "source");
	if (source == NULL)
		goto no_source;
	g_print ("source OK!\n");

	g_print ("Trying to create sink ... ");
	sink = gst_element_factory_make("autovideosink", "sink");
	if (sink == NULL)
		goto no_sink;
	g_print ("sink OK!\n");

	g_print ("Add elements to a bin ... \n");
	gst_bin_add_many(GST_BIN(pipeline), source, sink, NULL);

	g_print ("and link them ... \n");
	if (!gst_element_link_many(source, sink, NULL))
		g_warning("Failed to link elements!");

	bus = gst_pipeline_get_bus(GST_PIPELINE(pipeline));
	bus_watch_id = gst_bus_add_watch (bus, my_bus_callback, NULL);
	gst_object_unref(bus);

	g_print ("Play the pipeline ... \n");
	gst_element_set_state(pipeline, GST_STATE_PLAYING);

	g_print ("Start g_main_loop ... \n");
	loop = g_main_loop_new(NULL, FALSE);
	g_main_loop_run(loop);

	g_print ("Stop the pipeline ... \n");
	gst_element_set_state(pipeline, GST_STATE_NULL);

	gst_object_unref (pipeline);  // urefs source and sink?

	g_main_loop_unref(loop);

	return;

	no_source:
	{
		g_error ("Could not create GStreamer source element. "
				"Please install it");
		/* not reached, g_error aborts */
		return;
	}

	no_sink:
	{
		g_error ("Could not create GStreamer sink element. "
				"Please install it");
		/* not reached, g_error aborts */
		return;
	}
}

void
play_uri (const gchar * uri)
{
  GstStateChangeReturn sret;
  GstElement *playbin;
  GstElement *audiosink;
  GstElement *videosink;
  GstMessage *msg = NULL;
  GstBus *bus;

  g_print ("Trying to play %s ...\n", uri);

  playbin = gst_element_factory_make ("playbin", "playbin");
  if (playbin == NULL)
    goto no_playbin;

  /* get playbin's bus - we'll watch it for messages */
  bus = gst_pipeline_get_bus (GST_PIPELINE (playbin));

  /* set audio sink */
  audiosink = gst_element_factory_make ("autoaudiosink", "audiosink");
  if (audiosink == NULL)
    goto no_autoaudiosink;
  g_object_set (playbin, "audio-sink", audiosink, NULL);

  /* set video sink */
  videosink = gst_element_factory_make ("autovideosink", "videosink");
  if (videosink == NULL)
    goto no_autovideosink;
  g_object_set (playbin, "video-sink", videosink, NULL);

  /* set URI to play back */
  g_object_set (playbin, "uri", uri, NULL);

  /* and GO GO GO! */
  gst_element_set_state (GST_ELEMENT (playbin), GST_STATE_PLAYING);

  /* wait (blocks!) until state change either completes or fails */
  sret = gst_element_get_state (GST_ELEMENT (playbin), NULL, NULL, -1);

  switch (sret) {
    case GST_STATE_CHANGE_FAILURE:{
      msg = gst_bus_poll (bus, GST_MESSAGE_ERROR, 0);
      goto got_error_message;
    }
    case GST_STATE_CHANGE_SUCCESS:{
      GstMessage *msg;

      g_print ("Playing ...\n");

      while (1) {
        gint64 dur, pos;

        if (gst_element_query_duration (playbin, GST_FORMAT_TIME, &dur) &&
            gst_element_query_position (playbin, GST_FORMAT_TIME, &pos)) {
          g_print ("  %" GST_TIME_FORMAT " / %" GST_TIME_FORMAT "\n",
              GST_TIME_ARGS (pos), GST_TIME_ARGS (dur));
        }

        /* check if we finished or if there was an error,
         * but don't wait/block if neither is the case */
        msg = gst_bus_poll (bus, GST_MESSAGE_EOS | GST_MESSAGE_ERROR, 0);

        if (msg && GST_MESSAGE_TYPE (msg) == GST_MESSAGE_ERROR)
          goto got_error_message;

        if (msg && GST_MESSAGE_TYPE (msg) == GST_MESSAGE_EOS) {
          g_print ("Finished.\n");
          break;
        }

        /* sleep for one second */
        g_usleep (G_USEC_PER_SEC * 1);
      }
      break;
    }
    default:
      g_assert_not_reached ();
      break;
  }

  /* shut down and free everything */
  gst_element_set_state (playbin, GST_STATE_NULL);
  gst_object_unref (playbin);
  gst_object_unref (bus);
  return;

/* ERRORS */
got_error_message:
  {
    if (msg) {
      GError *err = NULL;
      gchar *dbg_str = NULL;

      gst_message_parse_error (msg, &err, &dbg_str);
      g_printerr ("FAILED to play %s: %s\n%s\n", uri, err->message,
          (dbg_str) ? dbg_str : "(no debugging information)");
      g_error_free (err);
      g_free (dbg_str);
      gst_message_unref (msg);
    } else {
      g_printerr ("FAILED to play %s: unknown error\n", uri);
    }

    /* shut down and free everything */
    gst_element_set_state (playbin, GST_STATE_NULL);
    gst_object_unref (playbin);
    gst_object_unref (bus);
    return;
  }

no_playbin:
  {
    g_error ("Could not create GStreamer 'playbin' element. "
        "Please install it");
    /* not reached, g_error aborts */
    return;
  }

no_autoaudiosink:
  {
    g_error ("Could not create GStreamer 'autoaudiosink' element. "
        "Please install it");
    /* not reached, g_error aborts */
    return;
  }

no_autovideosink:
  {
    g_error ("Could not create GStreamer 'autovideosink' element. "
        "Please install it");
    /* not reached, g_error aborts */
    return;
  }
}


